### ISSUE TITLE
InvalidProgramException at calling static method with AdviceArgumentSource.Instance attribute

### ISSUE THREAD

**Issue by kazachis4e** [pamidur/aspect-injector#21]
```C#
public class MyAspect
{
    [Advice(InjectionPoints.Before, InjectionTargets.Method)]
    public void AspectMethodRealization(
        [AdviceArgument(AdviceArgumentSource.Instance)] object instance)
    {
        Console.WriteLine("Aspect Works!!!");
    }
}

[Aspect(typeof(MyAspect))]
public static class MyStatic
{
    public static void StaticTest()
    {
        Console.WriteLine("Static Test");
    }
}

class Program
{
    static void Main(string[] args)
    {
        MyStatic.StaticTest();
    }
}
```

System.InvalidProgramException with message "Common Language Runtime detected an invalid program." appears at 'MyStatic.StaticTest();' line. If remove parameter 'object instance' from AspectMethodRealization() - all good.

In my case i need to read CustomAttributes of target method. Also i need to detect IsPublic or IsPrivate properties of target method (method can be static or non-static, does not matter for me). In previous realization (PostSharp) i do it by reflection (a have MethodBase object for target method), but now a have only TargetName property. So i need to correct Instanse property for getting MathodBase property. Or is there some another way to do it?

**Comment by pamidur** [comment]
Thanks for the report. Although I can reproduce this issue I don't have a full picture of the way it should be fixed.
So if you have instance of non-static class - you get it via AdviceArgumentSource.Instance
But What is supposed to be returned in case of static class? Null? Exception?
typeof(non-static-class) would a bit inconsistent here. wouldn't it?
Any thoughts ?

**Comment by kazachis4e** [comment]
I think in static case Instance must return null, not Exception. But in this case too little information about static method. So in my opinion there is two options:
    1. Add new AdviceArgumentSource property, e.g. TargetMethod, with MathodBase (or MethodInfo) type. If aspect marked with InjectionTargets.Method (and may be  InjectionTargets.Constructor) - fill this property. In another InjectionTargets - this property is null.
    2. Replace AdviceArgumentSource.TargetName to AdviceArgumentSource.MemberInfo with MemberInfo type. In aspect body MemberInfo can be casted to Methodinfo or PropertyInfo according InjectionTargets. Removed TargetName can be getted from MemberInfo.

**Comment by pamidur** [comment]
The point is that retrieving MemberInfo has huge performance impact because of runtime reflection. And TargetName is well known in compile time. May be we could leave both TargetName and TargetMember. So in case of property TargetName would be "SomeProperty" and TargetMember would be memberinfo of set_SomeProperty. 
Though there is one requirement - stick with portable subset of .net

**Comment by pamidur** [comment]
Fixed in 0.9.44.
I made it return null when member is static and instance is requested.
I also added new AdviceArgumentSource.Type so can access target type now. I guess you can handle retrieving member-infos with AdviceArgumentSource.Type + AdviceArgumentSource.TargetName for now.

I'll also create an improvement issue on backlog about something like AdviceArgumentSource.TargetMember. I hope we will find a way to make it fast and efficient.

**Comment by kazachis4e** [comment]
Very big thanks!!! Now this project completely satisfied my requirements )
